qadence/states.py

```diff
@@ -201,10 +205,15 @@ def product_state(
     ```python exec="on" source="material-block" result="json"
     from qadence.states import product_state
 
-    print(product_state("1100"))
+    print(product_state("1100", backend="pyqtorch"))
+    print(product_state("1100", backend="horqrux"))
     ```
     """
-    return _state_from_bitstring(bitstring, batch_size, endianness=endianness)
+    _state = [run(product_block(bitstring), backend=backend, endianness=endianness) for _ in range(batch_size)]
+    if backend in ["pyqtorch", "braket"]:
+        return torch.cat(_state)
+    elif backend in ["horqrux"]:
```

Comment:
we have other backends which are not open-source at the moment and which will automatically be available when installed. for simplicity's sake, let's assume that `product_state` will be called on a single `bitstring` and will therefor only return a single `ArrayLike`, so no need for `cat`

Reply:
That would render the `batch_size` argument useless. Are you in favour of removing it?

Reply:
Okay, does this mean I should return `jnp.array(_state)` or just `_state`?

I tried doing the first and I think casting a `torch.tensor` type state returned by some backends to `jnp.array()` is not a good idea. While testing it creates issues in managing the dimensions. Moreover, functions like `is_normalized` and `equivalent_state` used during testing are incompatible with `ArrayLike` inputs, which leads me to convert the arrays into tensors again just for the sake of testing.

Another thing I tried is to remove the `batch_size` argument entirely. Things work since we only have to deal with either `tensor` or `jnp.array`, however, I am unsure how important the `batch_size` argument is in the codebase. Let me know what you think.

Reply:
1. you can just return `_state`, each backend already returns the correct type (jax.Array or torch.Tensor)
2. `ArrayLike` is not the actual type, its more a Placeholder to say "its either a torch.Tensor or a jax.Array"
3. To test outputs or either JAX or torch against eachother, we use [jarr_to_tensor](https://github.com/pasqal-io/qadence/blob/main/tests/backends/test_backends.py#L274)
4. We can keep the interface of `product_state` the same however throw a `DeprecationWarning` if someone passes a `batch_size` arg, and finally add the `backend` arg last. There might be some tests which use it but we can rewrite them. Its something i have been wanting to change for a while

thanks a lot for helping us clean up the a bit messier parts of our codebase @SaashaJoshi, its really appreciated

Reply:
No problem. tbh I really liked your codebase, it is very clean!

Reply:
thank you :)